swift/Ice/async/Server/main.swift

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+import Foundation
+import Ice
+import PromiseKit
+
+class HelloI: Hello {
+    var workQueue: OperationQueue
+    var serialQueue: DispatchQueue // Used for blocking the change of workQueue.isSuspended
+    var concurrentQueue: DispatchQueue // DispatchQueue used in workQueue
+
+    init(_ queue: OperationQueue) {
+        workQueue = queue
+        serialQueue = DispatchQueue(label: "com.zeroc.demo.ice.async.server.serialqueue")
+        concurrentQueue = DispatchQueue(label: "com.zeroc.demo.ice.async.server.workqueue",
+                                        qos: .default,
+                                        attributes: .concurrent)
+        workQueue.underlyingQueue = concurrentQueue
+    }
+
+    func sayHelloAsync(delay: Int32, current _: Current) -> Promise<Void> {
+        return Promise { seal in
+            if delay == 0 {
+                print("Hello World!")
+                seal.fulfill(())
+            } else {
+                serialQueue.sync {
+                    guard self.workQueue.isSuspended == false else {
+                        seal.reject(RequestCanceledException())
+                        return
+                    }
+                    self.workQueue.addOperation {
+                        Thread.sleep(forTimeInterval: TimeInterval(delay) / 1000)
+                        print("Belated Hello World!")
+                        seal.fulfill(())
+                    }
+                }
+            }
+        }
+    }
+
+    func shutdown(current: Current) throws {
+        print("Shutting down...")
+        // Setting this property to true prevents the queue from starting any queued operations,
+        // but already executing operations continue to execute.
+        serialQueue.sync {
```

Comment:
We can dispatch this in the `concurrentQueue` with a barrier and get rid of the `serialQueue`